src/ipc/ipc4/helper.c

```diff
@@ -658,6 +659,74 @@ static int construct_config(struct ipc4_copier_module_cfg *copier_cfg, uint32_t
 	return IPC4_SUCCESS;
 }
 
+#if CONFIG_COMP_CHAIN_DMA
+int ipc4_create_mng_for_chain(struct ipc4_chain_dma *cdma)
+{
+	const struct sof_uuid uuid = {0x6a0a274f, 0x27cc, 0x4afb, {0xa3, 0xe7, 0x34,
+			0x44, 0x72, 0x3f, 0x43, 0x2e}};
+
+	const struct comp_driver *drv;
+	struct comp_dev *dev;
+
+	drv = ipc4_get_drv((uint8_t *)&uuid);
+	if (!drv)
+		return IPC4_INVALID_RESOURCE_ID;
+
+	dev = drv->ops.create(drv, (const struct comp_ipc_config *)NULL,
+			      (const void *)cdma);
+	if (!dev)
+		return IPC4_INVALID_RESOURCE_ID;
+
+	const uint32_t comp_id = IPC4_COMP_ID(cdma->primary.r.host_dma_id
+					      + IPC4_MAX_MODULE_COUNT, 0);
+	dev->ipc_config.id = comp_id;
+
+	list_init(&dev->bsource_list);
+	list_init(&dev->bsink_list);
+
+	return ipc4_add_comp_dev(dev);
+}
+
+int ipc4_chain_dma_state(struct comp_dev *dev, struct ipc4_chain_dma *cdma)
+{
+	const uint8_t link_dma_id = cdma->primary.r.link_dma_id;
+	const uint8_t host_dma_id = cdma->primary.r.host_dma_id;
+	const bool allocate = cdma->primary.r.allocate;
+	const bool enable = cdma->primary.r.enable;
+	int ret;
+
+	if (!dev)
+		return IPC4_FAILURE;
+
+	const struct comp_driver *drv = dev->drv;
+
+	uint32_t dir, host_chan, link_chan;
+
+	ret = process_dma_index(host_dma_id, &dir, &host_chan);
+
+	if (ret != IPC4_SUCCESS)
+		return ret;
+
+	ret = process_dma_index(link_dma_id, &dir, &link_chan);
+	if (ret != IPC4_SUCCESS)
+		return ret;
+
+	if (allocate) {
+		if (enable)
+			ret = drv->ops.trigger(dev, COMP_TRIGGER_START);
+		else
+			ret = drv->ops.trigger(dev, COMP_TRIGGER_PAUSE);
+	} else {
+		if (enable)
+			return IPC4_INVALID_REQUEST;
+		ret = drv->ops.trigger(dev, COMP_TRIGGER_PAUSE);
+		if (ret == IPC4_SUCCESS)
```

Comment:
Is this condition right? Not `if (ret != IPC4_SUCCESS)` ?

Reply:
I see, this is the stop and release the ChainDMA and only free if the trigger was succesfull?

Reply:
I meant that if 'trigger pause' return any error, remove flow stops and return error without removing chain. I could transform it into more readible form:

```suggestion
		/* remove chain */
		ret = drv->ops.trigger(dev, COMP_TRIGGER_PAUSE);
		if (ret)
		        return ret;
		drv->ops.free(dev);
```
Does it look better ?

Reply:
> 

yes, exactly

Reply:
> I see, this is the stop and release the ChainDMA and only free if the trigger was succesfull?



> I see, this is the stop and release the ChainDMA and only free if the trigger was succesfull?

yes, exactly